README.md

```diff
@@ -1 +1,267 @@
+>
+> — [DAG-JSON Spec](https://ipld.io/specs/codecs/dag-json/spec/)
+
+As it is critical for guarding against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+### 2.1.1 Caching & Invalidation
+
+Signing CIDs has two additional caching effects
+
+* Signing CIDs enables a simple strategy for caching validation by CID
+* Such a strategy also requires accounting for revocation of the signing keys, and so need to be tagged with this additional metadata
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `0x55`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+After being decoded from Varints, a varsig includes the following segments:
+
+| Segment Name              | Type          | Description                                     | Required |
+|---------------------------|---------------|-------------------------------------------------|----------|
+| Varsig Multiformat Prefix | `0x34`        | The multicodec varsig prefix                    | Yes      |
+| Content Multicodec Prefix | `Multiformat` | The IPLD encoding used to canonicalize the data | Yes      |
+| Multihash                 | `Multihash`   | The multicodec prefix for the multihash         | Yes      |
+| Public Key Prefix         | `Multicodec`  | The multicodec prefix for the public key type   | Yes      |
+| Raw Signature             | `Bytes`       | The raw signature                               | Yes      |
+  
+## 3.1 Segments
+
+### 3.1.1 Varsig Prefix
+
+The varsig prefix MUST be `0x34`.
+
+### 3.1.2 Hash Prefix
+
+The [multicodec](https://github.com/multiformats/multicodec) for the [multihash](https://github.com/multiformats/multihash) used by the signature. This is the first segment of a .
+
+For example: `0x16` for SHA3-256, or `0x1e` for BLAKE3
+
+### 3.1.3 Public Key Prefix
+
+The [multicodec](https://github.com/multiformats/multicodec) prefix for the public key type associated with the signature.
+
+For example: `0xe7` for secp256k1, or `0x1205` for RSA.
+
+### 3.1.4 Content Multicodec Prefix
+
+The [multicodec](https://github.com/multiformats/multicodec) prefix of the encoding scheme used.
+
+For example: `0x55` for raw bytes (no special encoding), `0x0129` for DAG-JSON, and `0x70` for DAG-PB.
+
+### 3.1.5 Raw Signature
+
+The raw signature bytes
+
+## 3.2 Segment Layout
+
+```xml
+<multiformat 0x34><varint content_multicodec><varint multihash><varint key_multicodec><varint raw_signature>
```

Comment:
Still wondering about the choice of `<varint raw_signature>`.

Is this more or less space efficient than `<varint signature_length><raw_signature>`?

Reply:
I think this is a typo I think @expede meant to put `<bytes raw_signature>` instead. Unless I'm mistaken we no longer need `signature_length` because `multihash` has it.

Reply:
@expede I'm bit confused about this, because multihash provides hash alg, digest size and actual digest bytes in it

![](https://raw.githubusercontent.com/multiformats/multihash/master/img/multihash.006.jpg)

Here we embed multihash with a digest and then also embed signature bytes ? Or maybe I've misunderstood and you meant to say `<varint hash-fn>` instead of `<varint multihash>` ? If so I think we need to also need to encode `signature_length`  as @oed pointed out above because multiformats should be embeddable and therefor one should be able to tell where they end. Multihash itself states following about why `hash_size` is required

> Because otherwise you end up with a function code really meaning "function-and-digest-size-code". Makes using custom digest sizes annoying, and is less flexible.